File: attocode/src/integrations/swarm/request-throttle.ts
/**
 * Swarm Request Throttle
 *
 * Token bucket + minimum spacing + FIFO queue to prevent 429 rate limiting
 * across all swarm workers. Since all subagents share the parent's provider
 * by reference (agent.ts:4398), wrapping the provider here automatically
 * throttles ALL downstream LLM calls.
 */

import type { LLMProvider, Message, ChatOptions, ChatResponse, MessageWithContent } from '../../providers/types.js';

// ─── Configuration ──────────────────────────────────────────────────────────

export interface ThrottleConfig {
  /** Burst capacity — max concurrent in-flight requests (default: 2) */
  maxConcurrent: number;
  /** Token refill rate per second (default: 0.5 → 30 req/min) */
  refillRatePerSecond: number;
  /** Floor between consecutive request starts in ms (default: 2000) */
  minSpacingMs: number;
}

export const FREE_TIER_THROTTLE: ThrottleConfig = {
  maxConcurrent: 2,
  refillRatePerSecond: 0.5,
  minSpacingMs: 1500,
};

export const PAID_TIER_THROTTLE: ThrottleConfig = {
  maxConcurrent: 5,
  refillRatePerSecond: 2.0,
  minSpacingMs: 200,
};

// ─── Throttle Stats ─────────────────────────────────────────────────────────

export interface ThrottleStats {
  pendingCount: number;
  availableTokens: number;
  totalAcquired: number;
  backoffLevel: number;
  currentMaxConcurrent: number;
  currentMinSpacingMs: number;
}

// ─── SwarmThrottle ──────────────────────────────────────────────────────────

/**
 * Async semaphore with token bucket rate limiting and minimum spacing.
 *
 * - `acquire()` waits in FIFO order, then consumes a token once available.
 * - Tokens refill passively based on elapsed wall-clock time.
 * - No explicit `release()` — long LLM calls naturally allow refill.
 */
export class SwarmThrottle {
  private tokens: number;
  private lastAcquireTime = 0;
  private lastRefillTime: number;
  private queue: Array<() => void> = [];
  private totalAcquired = 0;
  private refillTimer: ReturnType<typeof setInterval> | null = null;

  // Adaptive backoff state
  private originalConfig: Readonly<ThrottleConfig>;
  private _backoffLevel = 0;
  private lastBackoffTime = 0;
  private static readonly MAX_BACKOFF_LEVEL = 3;
  private static readonly RECOVER_COOLDOWN_MS = 10_000;

  constructor(private config: ThrottleConfig) {
    this.tokens = config.maxConcurrent;
    this.lastRefillTime = Date.now();
    this.originalConfig = { ...config };
  }

  /**
   * Wait for a token to become available (FIFO).
   * Enforces minSpacing between consecutive acquisitions.
   */
  async acquire(): Promise<void> {
    // FIFO: wait in queue if there are already waiters or no tokens
    if (this.queue.length > 0 || this.refill() < 1) {
      await new Promise<void>(resolve => {
        this.queue.push(resolve);
        this.ensureRefillTimer();
      });
    }

    // Enforce minimum spacing
    const now = Date.now();
    const elapsed = now - this.lastAcquireTime;
    if (elapsed < this.config.minSpacingMs && this.lastAcquireTime > 0) {
      const waitMs = this.config.minSpacingMs - elapsed;
      await sleep(waitMs);
    }

    // Consume token (refill first after possible wait)
    this.refill();
    this.tokens = Math.max(0, this.tokens - 1);
    this.lastAcquireTime = Date.now();
    this.totalAcquired++;
  }

  /**
   * Get current throttle stats for observability.
   */
  getStats(): ThrottleStats {
    this.refill();
    return {
      pendingCount: this.queue.length,
      availableTokens: Math.floor(this.tokens),
      totalAcquired: this.totalAcquired,
      backoffLevel: this._backoffLevel,
      currentMaxConcurrent: this.config.maxConcurrent,
      currentMinSpacingMs: this.config.minSpacingMs,
    };
  }

  /** Current backoff level (0 = normal, max = 3). */
  get backoffLevel(): number {
    return this._backoffLevel;
  }

  /**
   * Increase backoff: halve maxConcurrent (min 1), double minSpacingMs (max 5000),
   * halve refillRate (min 0.1). Caps at level 3.
   */
  backoff(): void {
    if (this._backoffLevel >= SwarmThrottle.MAX_BACKOFF_LEVEL) return;
    this._backoffLevel++;
    this.lastBackoffTime = Date.now();

    this.config = {
      maxConcurrent: Math.max(1, Math.floor(this.config.maxConcurrent / 2)),
      minSpacingMs: Math.min(5000, this.config.minSpacingMs * 2),
      refillRatePerSecond: Math.max(0.1, this.config.refillRatePerSecond / 2),
    };

    // Clamp tokens to new maxConcurrent
    this.tokens = Math.min(this.tokens, this.config.maxConcurrent);
  }

  /**
   * Step back toward original config after sustained success.
   * Only recovers if 10s+ have passed since last backoff.
   */
  recover(): void {
    if (this._backoffLevel <= 0) return;
    if (Date.now() - this.lastBackoffTime < SwarmThrottle.RECOVER_COOLDOWN_MS) return;

    this._backoffLevel--;

    // Interpolate back toward original config
    const ratio = this._backoffLevel / SwarmThrottle.MAX_BACKOFF_LEVEL;
    this.config = {
      maxConcurrent: Math.max(1, Math.round(
        this.originalConfig.maxConcurrent * (1 - ratio * 0.5),
      )),
      minSpacingMs: Math.round(
        this.originalConfig.minSpacingMs * (1 + ratio),
      ),
      refillRatePerSecond: this.originalConfig.refillRatePerSecond * (1 - ratio * 0.5),
    };
  }

  /**
   * Feed rate limit info from response headers to proactively adjust throttle.
   * If remaining requests or tokens are low, preemptively back off.
   */
  feedRateLimitInfo(info: { remainingRequests?: number; remainingTokens?: number; resetSeconds?: number }): void {
    // Proactive backoff: if < 5 remaining requests, increase spacing
    if (info.remainingRequests !== undefined && info.remainingRequests < 5) {
      if (this._backoffLevel < SwarmThrottle.MAX_BACKOFF_LEVEL) {
        this.backoff();
      }
    }
    // If reset is imminent (< 2s), briefly pause
    if (info.resetSeconds !== undefined && info.resetSeconds < 2 && info.remainingRequests !== undefined && info.remainingRequests <= 1) {
      this.config = {
        ...this.config,
        minSpacingMs: Math.max(this.config.minSpacingMs, info.resetSeconds * 1000),
      };
    }
  }

  /**
   * Refill tokens based on elapsed time. Returns current token count.
   */
  private refill(): number {
    const now = Date.now();
    const elapsedSec = (now - this.lastRefillTime) / 1000;
    if (elapsedSec > 0) {
      this.tokens = Math.min(
        this.config.maxConcurrent,
        this.tokens + elapsedSec * this.config.refillRatePerSecond,
      );
      this.lastRefillTime = now;
    }
    return this.tokens;
  }

  /**
   * Start a timer to periodically check for token refills and wake waiters.
   * Automatically stops when the queue is empty.
   */
  private ensureRefillTimer(): void {
    if (this.refillTimer !== null) return;

    // Check every ~50ms or minSpacingMs, whichever is smaller
    const intervalMs = Math.min(50, this.config.minSpacingMs || 50);
    this.refillTimer = setInterval(() => {
      this.refill();
      // Wake queued waiters if tokens available
      while (this.queue.length > 0 && this.tokens >= 1) {
        const resolve = this.queue.shift();
        if (resolve) resolve();
      }
      // Stop timer when queue is drained
      if (this.queue.length === 0 && this.refillTimer !== null) {
        clearInterval(this.refillTimer);
        this.refillTimer = null;
      }
    }, intervalMs);
  }
}

// ─── ThrottledProvider ──────────────────────────────────────────────────────

/**
 * Wraps an LLMProvider with throttle.acquire() before each chat call.
 * Delegates name, defaultModel, isConfigured() to inner provider.
 */
export class ThrottledProvider implements LLMProvider {
  constructor(
    private inner: LLMProvider,
    private throttle: SwarmThrottle,
  ) {}

  get name(): string {
    return this.inner.name;
  }

  get defaultModel(): string {
    return this.inner.defaultModel;
  }

  isConfigured(): boolean {
    return this.inner.isConfigured();
  }

  async chat(messages: (Message | MessageWithContent)[], options?: ChatOptions): Promise<ChatResponse> {
    await this.throttle.acquire();
    try {
      const result = await this.inner.chat(messages, options);
      // V3: Feed rate limit info from response headers for proactive throttle
      if (result.rateLimitInfo) {
        this.throttle.feedRateLimitInfo(result.rateLimitInfo);
      }
      this.throttle.recover();
      return result;
    } catch (error) {
      if (isRateLimitError(error)) {
        this.throttle.backoff();
      }
      throw error;
    }
  }

  /**
   * Forward chatWithTools if the inner provider supports it.
   */
  async chatWithTools(...args: unknown[]): Promise<unknown> {
    await this.throttle.acquire();
    const inner = this.inner as unknown as Record<string, (...a: unknown[]) => unknown>;
    if (typeof inner.chatWithTools === 'function') {
      try {
        const result = await inner.chatWithTools(...args);
        this.throttle.recover();
        return result;
      } catch (error) {
        if (isRateLimitError(error)) {
          this.throttle.backoff();
        }
        throw error;
      }
    }
    throw new Error('Inner provider does not support chatWithTools');
  }

  /** Get the underlying throttle for stats/inspection. */
  getThrottle(): SwarmThrottle {
    return this.throttle;
  }
}

// ─── Factory ────────────────────────────────────────────────────────────────

/**
 * Create a throttled wrapper around an LLM provider.
 */
export function createThrottledProvider(
  provider: LLMProvider,
  config: ThrottleConfig = FREE_TIER_THROTTLE,
): ThrottledProvider {
  const throttle = new SwarmThrottle(config);
  return new ThrottledProvider(provider, throttle);
}

// ─── Helpers ────────────────────────────────────────────────────────────────

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/** Check if an error is a rate limit (429) or spend limit (402) error. */
function isRateLimitError(error: unknown): boolean {
  const msg = String(error instanceof Error ? error.message : error).toLowerCase();
  return msg.includes('429') || msg.includes('rate') || msg.includes('too many') || msg.includes('402');
}

File: attocode/attocode_py/src/attocode/integrations/swarm/request_throttle.py
"""Request throttle for rate limiting swarm workers.

Token bucket + minimum spacing + FIFO queue to prevent 429 rate limiting
across all swarm workers. Wraps LLM providers to throttle downstream calls.
"""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass, field
from typing import Any

from attocode.providers.base import LLMProvider


@dataclass(slots=True)
class ThrottleConfig:
    """Rate limiting configuration."""

    max_concurrent: int = 5
    refill_rate_per_second: float = 2.0
    min_spacing_ms: float = 200.0
    max_backoff_level: int = 3


# Preset configs
FREE_TIER_THROTTLE = ThrottleConfig(
    max_concurrent=2,
    refill_rate_per_second=0.5,
    min_spacing_ms=1500.0,
)

PAID_TIER_THROTTLE = ThrottleConfig(
    max_concurrent=5,
    refill_rate_per_second=2.0,
    min_spacing_ms=200.0,
)


@dataclass(slots=True)
class ThrottleStats:
    """Observable throttle statistics."""

    pending_count: int = 0
    available_tokens: int = 0
    total_acquired: int = 0
    backoff_level: int = 0
    current_max_concurrent: int = 0
    current_min_spacing_ms: float = 0.0


class SwarmThrottle:
    """Token bucket rate limiter with FIFO queue and adaptive backoff.

    Features:
    - Token bucket with configurable refill rate
    - Minimum spacing between requests
    - FIFO ordering for fairness
    - Adaptive backoff on 429 errors
    - Automatic recovery after sustained success
    """

    def __init__(self, config: ThrottleConfig | None = None) -> None:
        self._config = config or ThrottleConfig()
        self._semaphore = asyncio.Semaphore(self._config.max_concurrent)
        self._last_request_time = 0.0
        self._lock = asyncio.Lock()
        self._total_acquired = 0
        self._pending = 0

        # Adaptive state
        self._backoff_level = 0
        self._current_max_concurrent = self._config.max_concurrent
        self._current_min_spacing_ms = self._config.min_spacing_ms
        self._current_refill_rate = self._config.refill_rate_per_second
        self._last_backoff_time = 0.0
        self._success_since_backoff = 0

    async def acquire(self) -> None:
        """FIFO wait for a throttle token with minimum spacing."""
        self._pending += 1
        try:
            await self._semaphore.acquire()

            # Enforce minimum spacing
            async with self._lock:
                now = time.monotonic()
                spacing_sec = self._current_min_spacing_ms / 1000.0
                elapsed = now - self._last_request_time
                if elapsed < spacing_sec:
                    await asyncio.sleep(spacing_sec - elapsed)
                self._last_request_time = time.monotonic()
                self._total_acquired += 1
        finally:
            self._pending -= 1

    def release(self) -> None:
        """Release a throttle token."""
        self._semaphore.release()

    def backoff(self) -> None:
        """Increase throttling after a rate limit error."""
        if self._backoff_level >= self._config.max_backoff_level:
            return

        self._backoff_level += 1
        self._last_backoff_time = time.monotonic()
        self._success_since_backoff = 0

        # Halve concurrency, double spacing, halve refill
        self._current_max_concurrent = max(1, self._config.max_concurrent >> self._backoff_level)
        self._current_min_spacing_ms = self._config.min_spacing_ms * (2 ** self._backoff_level)
        self._current_refill_rate = self._config.refill_rate_per_second / (2 ** self._backoff_level)

    def recover(self) -> None:
        """Step back toward original config after sustained success."""
        if self._backoff_level <= 0:
            return

        # Only recover if we've had 10+ successes since last backoff
        self._success_since_backoff += 1
        if self._success_since_backoff < 10:
            return

        # And at least 10 seconds have passed
        if time.monotonic() - self._last_backoff_time < 10.0:
            return

        self._backoff_level -= 1
        self._success_since_backoff = 0

        if self._backoff_level <= 0:
            self._current_max_concurrent = self._config.max_concurrent
            self._current_min_spacing_ms = self._config.min_spacing_ms
            self._current_refill_rate = self._config.refill_rate_per_second
        else:
            self._current_max_concurrent = max(1, self._config.max_concurrent >> self._backoff_level)
            self._current_min_spacing_ms = self._config.min_spacing_ms * (2 ** self._backoff_level)
            self._current_refill_rate = self._config.refill_rate_per_second / (2 ** self._backoff_level)

    def feed_rate_limit_info(self, remaining: int | None = None, reset_ms: int | None = None) -> None:
        """Proactive adjustment from response headers."""
        if remaining is not None and remaining <= 1:
            self.backoff()
        if reset_ms is not None and reset_ms > 5000:
            self.backoff()

    def get_stats(self) -> ThrottleStats:
        """Get current throttle statistics."""
        return ThrottleStats(
            pending_count=self._pending,
            available_tokens=self._current_max_concurrent,
            total_acquired=self._total_acquired,
            backoff_level=self._backoff_level,
            current_max_concurrent=self._current_max_concurrent,
            current_min_spacing_ms=self._current_min_spacing_ms,
        )


class ThrottledProvider:
    """Wraps an LLM provider with rate throttling.

    Acquires throttle token before each API call, recovers on success,
    backs off on rate limit (429) or spend limit (402) errors.
    """

    def __init__(self, provider: LLMProvider, throttle: SwarmThrottle) -> None:
        self._provider = provider
        self._throttle = throttle

    async def chat(self, messages: list[Any], **kwargs: Any) -> Any:
        """Throttled chat call."""
        await self._throttle.acquire()
        try:
            result = await self._provider.chat(messages, **kwargs)
            self._throttle.recover()
            return result
        except Exception as e:
            error_str = str(e).lower()
            if "429" in error_str or "rate" in error_str:
                self._throttle.backoff()
            elif "402" in error_str:
                self._throttle.backoff()
            raise
        finally:
            self._throttle.release()

    def get_throttle(self) -> SwarmThrottle:
        """Access underlying throttle for inspection."""
        return self._throttle


def create_throttled_provider(
    provider: LLMProvider,
    config: ThrottleConfig | None = None,
) -> ThrottledProvider:
    """Factory function to create a throttled provider."""
    throttle = SwarmThrottle(config)
    return ThrottledProvider(provider, throttle)
